x-pack/plugins/files/server/blob_storage_service/adapters/es/es.test.ts

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+import { Readable } from 'stream';
+import { promisify } from 'util';
+import { ElasticsearchClient } from '@kbn/core-elasticsearch-server';
+import { loggingSystemMock } from '@kbn/core-logging-server-mocks';
+import { elasticsearchServiceMock } from '@kbn/core/server/mocks';
+import { Semaphore } from '@kbn/semaphore';
+import * as sinon from 'sinon';
+
+import { ElasticsearchBlobStorageClient } from './es';
+
+const setImmediate = promisify(global.setImmediate);
+
+describe('ElasticsearchBlobStorageClient', () => {
+  let esClient: ElasticsearchClient;
+  let blobStoreClient: ElasticsearchBlobStorageClient;
+  let semaphore: Semaphore;
+
+  beforeEach(() => {
+    semaphore = new Semaphore(1);
+    esClient = elasticsearchServiceMock.createElasticsearchClient();
+    blobStoreClient = new ElasticsearchBlobStorageClient(
+      esClient,
+      undefined,
+      undefined,
+      loggingSystemMock.createLogger(),
+      semaphore
+    );
+  });
+
+  test('limits max concurrent uploads', async () => {
+    const acquireSpy = sinon.spy(semaphore, 'acquire');
+    (esClient.index as jest.Mock).mockImplementation(() => {
+      return new Promise((res, rej) => setTimeout(() => rej('failed'), 100));
+    });
+    const [p1, p2, ...rest] = [
+      blobStoreClient.upload(Readable.from(['test'])).catch(() => {}),
+      blobStoreClient.upload(Readable.from(['test'])).catch(() => {}),
+      blobStoreClient.upload(Readable.from(['test'])).catch(() => {}),
+      blobStoreClient.upload(Readable.from(['test'])).catch(() => {}),
+    ];
+    await setImmediate();
+    expect(acquireSpy.callCount).toBe(4);
```

Comment:
If we use `jest.spyOn`:
```suggestion
    expect(acquireSpy).toHaveBeenCalledTimes(4);
```